cpp/src/arrow/compute/exec/spilling_join.cc

```diff
@@ -0,0 +1,351 @@
+        {
+            size_t size_of_batch = static_cast<size_t>(batch.TotalBufferSize());
+            size_t max_batch_size = arrow::util::AtomicMax(max_batch_size_, size_of_batch);
+
+            // Spilling algorithm proven to not use more than
+            // (SpillThreshold + NumThreads * BatchSize) memory.
+            // Thus we want to spill when (SpillThreshold + NumThreads * BatchSize) = k * MaxMemory
+            // with some fuzz factor k (which is 0.8 here because that's what I decided). 
+            // Thus SpillThreshold = k * MaxMemory - NumThreads * BatchSize. 
+            constexpr float kFuzzFactor = 0.8f;
+            size_t max_memory = static_cast<size_t>(kFuzzFactor * ctx_->options().max_memory_bytes);
+            size_t spill_threshold =
+                static_cast<size_t>(
+                    std::max(
+                        static_cast<ssize_t>(kFuzzFactor * max_memory - num_threads_ * max_batch_size),
+                        static_cast<ssize_t>(0)));
+            size_t bytes_allocated = static_cast<size_t>(ctx_->memory_pool()->bytes_allocated());
+            size_t bytes_inflight = ctx_->GetCurrentTempFileIO();
+
+            size_t backpressure_threshold = spill_threshold / 2;
+            if(bytes_allocated > backpressure_threshold)
+                callbacks_.pause_probe_side(backpressure_counter_.fetch_add(1));
+            if((bytes_allocated - bytes_inflight) > spill_threshold)
+                RETURN_NOT_OK(AdvanceSpillCursor(thread_index));
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::AdvanceSpillCursor(size_t thread_index)
+        {
+            bool expected = false;
+            if(!spilling_.load() && spilling_.compare_exchange_strong(expected, true))
+                return callbacks_.start_spilling(thread_index);
+
+            ARROW_ASSIGN_OR_RAISE(bool probe_advanced, probe_accumulator_.AdvanceSpillCursor());
+            if(probe_advanced) return Status::OK();
+
+            ARROW_ASSIGN_OR_RAISE(bool build_advanced, build_accumulator_.AdvanceSpillCursor());
+            if(build_advanced) return Status::OK();
+
+            ARROW_ASSIGN_OR_RAISE(bool probe_hash_advanced, probe_accumulator_.AdvanceHashCursor());
+            if(probe_hash_advanced) return Status::OK();
+
+            ARROW_ASSIGN_OR_RAISE(bool build_hash_advanced, build_accumulator_.AdvanceHashCursor());
+            if(build_hash_advanced) return Status::OK();
+
+            // Pray we don't run out of memory
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::OnBuildSideBatch(size_t thread_index, ExecBatch batch)
+        {
+            return build_accumulator_.InsertBatch(
+                thread_index,
+                std::move(batch));
+        }
+
+        Status SpillingHashJoin::OnBuildSideFinished(size_t thread_index)
+        {
+            callbacks_.resume_probe_side(backpressure_counter_.fetch_add(1));
```

Comment:
resume_probe_side should happen when we have both the first hash table and the Bloom filter for all partitions not below build hash cursor ready. Resuming before building Bloom filter is too early.

Reply:
Moved it to when Bloom filter is finished.